kotlinx-coroutines-io/src/main/kotlin/kotlinx/coroutines/experimental/io/Coroutines.kt

```diff
@@ -128,9 +106,32 @@ fun writer(
     }
 }
 
-internal class ReaderCoroutine(context: CoroutineContext, channel: ByteChannel) :
-    ByteChannelCoroutine(context, channel), ReaderJob, ReaderScope
+/**
+ * @param S not exactly safe (unchecked cast is used) so should be [ReaderScope] or [WriterScope]
+ */
+private fun <S : CoroutineScope> CoroutineScope.launchChannel(context: CoroutineContext,
+                                                              channel: ByteChannel,
+                                                              attachJob: Boolean,
+                                                              block: suspend S.() -> Unit): ChannelJob {
+    val job = launch(context) {
+        if (attachJob) {
+            channel.attachJob(coroutineContext[Job]!!)
+        }
+        try {
+            @Suppress("UNCHECKED_CAST")
+            block(ChannelScope(this, channel) as S)
+        } catch (t: Throwable) {
+            channel.cancel(t)
+            throw CancellationException("Cancelled due to error $t")
```

Comment:
Why are you not using t as `cause`?

Reply:
To avoid double cancellation that caused strange effects in the past because of circular job dependency graph.  Not exactly sure about it

Reply:
Rewritten so exceptions are not caught at all but handled in `invokeOnCompletion`. Potential double cancellation handling is your problem from this point ;-)